Vmapper

marcmerlin opened the pull request:
Fixes https://github.com/hzeller/rpi-rgb-led-matrix/issues/1003

Ok, this was more work than I thought for something seemingly simple (well, it's likely just that I suck at it), but it works!
Hopefully this puts me in the lead for the highest usable resolution display on a single rPi  (384x256) :)
![401_20200310_FastLED_RPIRGBPanel_GFX_384x256](https://user-images.githubusercontent.com/1369412/76924580-05cd2380-6894-11ea-9a3d-4ace9b7fc710.jpg)

hzeller commented:
Cool!
Only potential user-support issue I see is that the panels are fed from the top while the 'usual' way is to feed them from the right.

So to reduce potential user-support in the future, I suggest to not deviate too much from the 'things are fed from the right side', and think about stuff rotated by 90 degrees (it is probably just a little x/y swap fiddle somewhere).

To make sure there is no confusion about how to connect each of these panels, can you add an text-art that indicates how they are exactly chained ? I find that the Macron-Character is excellent for this kind of stuff:
```
[v]¯\_[v]¯\_[v]¯\_[v]¯ ch1
[v]¯\_[v]¯\_[v]¯\_[v]¯ ch2
[v]¯\_[v]¯\_[v]¯\_[v]¯ ch3
```

(if you still want to operate it from the top in your specific application, you then can use the Rotate transformation to get it back rotated)

hzeller commented:
(also, this brings out much more of the `V`-ness)

marcmerlin commented:
Mmmh, so the whole point of vmapper is that the panels are fed up-down. Sure, I can rotate everything and then force the use of a second rotate:90 mapper to get it the way you're suggesting, but the whole point of V-mapper is that the panels are fed from up to down (or down to up).

I think you're saying that I can pretend that my panel is like this, is that what you meant?
![image](https://user-images.githubusercontent.com/1369412/76927369-fce05000-689b-11ea-9d7a-2fd81056da59.png)

If so, I'm not sure I'm not sure if your suggestion actually makes thing more obvious given that the point of V stacking is that you want matrix is naturally higher than wide, so it's not very obvious to rotate it 90 degrees.
That said, if somehow you have a usage case where you want it that way, you can always use rotate:90, but by default I think it ought to be the way I coded it, wouldn't you agree?
I'm happy to change the ASCII to show how the wiring is (as per your suggestion).
Would that sole change be ok?

hzeller commented:
mmh ok, so the reason why you want to feed it from the top is that then the panels are 'upright' as they are meant to be ? (because sometimes the LED shielding sideways is a little different than up/down). I could be convinced by that.

Let's do the ASCII art improvement for now, but since you turn it 90 degrees than in my suggestion, you probably need some sort of 'right side pipe symbol' and 'left side pipe symbol' and backslash to illustrate.

marcmerlin commented:
Just to answer your question, yes, the panels are upright as they were meant to be.
This is their natural position given the writing on the back, so I felt it made sense to keep them that way by default.
Of course if someone needs a resolution that is the other way around, they can use rotate.

hzeller commented on examples-api-use/README.md:
```diff
@@ -313,12 +313,15 @@ what you can reasonably output without having an unusable fresh rate
+  Ch1     Ch2     Ch3
+   v       v       v
+  [I > O] [I > O] [I > O]
+   ,---'   ,---'   ,---'
```
Oh, pretty :)

hzeller commented on lib/pixel-mapper.cc:
```diff
@@ -217,6 +217,54 @@ class UArrangementMapper : public PixelMapper {
+  virtual const char *GetName() const { return "V-mapper"; }
+
+  virtual bool SetParameters(int chain, int parallel, const char *param) {
+    if (chain < 2) {
```
Why this limitation ? With a chain of one it would be possible to make a long panel (e.g. for a runtext) that is fast to update (because it then would be three short segments, that make use of parallel feeding)

```
[O < I] [O < I] [O < I]
     |       |       |
     ch3     ch2      ch1
```

Looking at the code below, it _should_ be not a problem (we never divide by (chain_ - 1) for instance).

marcmerlin replied:
@hzeller  your are absolutely correct that chains of 1 would both work, and make sense. I updated that.
You are also correct that I got the ASCII art reverted (not counting left-right which is a bit tricky as it depends on whether you are showing the panels with pixels facing you or with the connectors facing you, but I think that one isn't too important, the up/down one is; that said I changed both).

marcmerlin commented:
@hzeller your are absolutely correct that chains of 1 would both work, and make sense. I updated that.
You are also correct that I got the ASCII art reverted (not counting left-right which is a bit tricky as it depends on whether you are showing the panels with pixels facing you or with the connectors facing you, but I think that one isn't too important, the up/down one is; that said I changed both).

hzeller commented:
Do you have a picture of the panel backside ? Every panel I have ever seen has the input on the right side (seeing from the face of the panel), not on the left.

hzeller commented:
Looking at the three-chained V-example you have above, I don't see arrows but the text is indeed up only if the input is on the right.

So looking to the front of the panel, the input is on the right, so we need to make sure to have the code reflect that.

marcmerlin commented:
![IMG_20200318_090455](https://user-images.githubusercontent.com/1369412/76981351-b3255300-68f7-11ea-99c4-4f5c00c31998.jpg)

hzeller commented:
See, input is on the right of the panel, output on the left
(looking from the back, the input is on the left, output on the right)

marcmerlin commented:
connector is left when seen from the back, right when seen from the LED (front) side.
The new graph I sent you is showing I on the left which is where it is when you are wiring the panels and the connectors are facing you.
Regardless of whether we show one side or the other, it should be clear enough to someone repeating this hopefully :)

hzeller commented:
The wiring diagrams in all of the rpi-rgb-led-matrix done are all with the panel fronts looking toward you.
So if you have this already wired and programmed this way, then just the graphics needs to be flipped.

marcmerlin commented:
Sure, PTAL

hzeller commented on examples-api-use/README.md:
```diff
@@ -291,6 +291,40 @@ two chains with 8 panels each
+   ,---'   ,---'   ,---'
+  [O < I] [O < I] [O < I]
+       ^       ^       ^
+      Ch1     Ch2     Ch3
```
After flipping, wouldn't now the ch1 be on the right ?

marcmerlin replied:
you are correct, sorry about that, fixed.

hzeller commented on examples-api-use/README.md:
````diff
@@ -291,6 +291,40 @@ two chains with 8 panels each
+```
+
+Note that this is not a U mapper, all the panels are correct side up, and you need
+more cable length. It looks like this, 3 chains of 4 panels for a total of 32K pixels
````
12 panels with 128x64 pixels each is actually a almost 100k pixels (98304)...

hzeller commented on examples-api-use/README.md:
```diff
@@ -291,6 +291,40 @@ two chains with 8 panels each
+
+Note that this is not a U mapper, all the panels are correct side up, and you need
+more cable length. It looks like this, 3 chains of 4 panels for a total of 32K pixels
+per chain. This is also a good time to mention that this is probably an upper limit of
```
... ah, just see the 'per chain :)' nvm.

hzeller commented on lib/pixel-mapper.cc:
```diff
@@ -217,6 +217,50 @@ class UArrangementMapper : public PixelMapper {
+    *matrix_x = (x % panel_width) +  int(y/panel_height)* panel_width;
+    *matrix_y = (y % panel_height) + int(x/panel_width) * panel_height;
+
+    //fprintf(stderr, "%s: Panel-W:%d Panel-H:%d. X: %3d -> %3d, Y: %3d -> %3d\n",
```
Don't leave commented-out code in the final version.

marcmerlin replied:
So, I left it there on purpose it case it helps someone (possibly me) writing the next mapper. Understanding how things worked without it would have been hard (it was non obvious to see how things were mapped)
If you'd prefer, I can put it under an #ifdef 0  or something else, though, but if you really don't want it, I can remove it and apply it again in my tree as I will definitely want it again in the future if I write another mapper.
What's your call?

hzeller replied:
Removing commented-out here. Someone who will want to debug this will add something very similar in the working copy, but here, we should leave out the clutter.

Also above in the GetSizeMapping(), the output should not be there, but you can `#if 0` that out if you want.

marcmerlin replied:
Ok, PTAL

hzeller commented:
Cool, almost ready to go: just some leftover debug printf to be removed.

hzeller commented:
Thanks for your contribution! Merged.

hzeller commented:
Do you have a picture of the back of your panel that you'd like to share ? I think similar to the U-mapper example image, that will help people in the wiring.

marcmerlin commented:
@hzeller I don't have rails, it's sadly all like just sitting on my desk, so I can't really flip it around
The picture I gave upthread of a single set of 3 panels is basically what a chain looks like (different panel, but identical wiring)

marcmerlin commented:
@hzeller  Well, I will owe you a new matter soon, Vumapper :)

![IMG_20200320_174205](https://user-images.githubusercontent.com/1369412/77215625-621d8680-6ad2-11ea-8fce-cbdf2649d330.jpg)
And I have the back of this one
![IMG_20200320_174436](https://user-images.githubusercontent.com/1369412/77215646-80838200-6ad2-11ea-977a-9d903933376c.jpg)

hzeller commented:
You're really going to town with these panels :)

marcmerlin commented:
those old 64x32 panels that were donated to me, were gathering dust, I felt bad about it, and like many I have a few extra hours at home right now :)
Don't worry, after a not too long while, I'll get bored with this, and switch to another hobby like RC planes :)

rijotech commented:
I'm trying to replicate the panels setup but can't seen to find the right set of commands to get a straight image. @marcmerlin care to provide an example? I already haverything wired the same way with 12 64x32 panels.

marcmerlin commented:
@rijotech please try
~/rpi-rgb-led-matrix/examples-api-use/demo --led-gpio-mapping=regular --led-rows=32 --led-cols=64  --led-row-addr-type=0 --led-chain=5 --led-show-refresh --led-slowdown-gpio=1 --led-parallel=3   --led-panel-type=FM6126A --led-pwm-bits=7 --led-pwm-dither-bits=1 --led-pixel-mapper=V-Mapper:Z -D0
If that fails, please post picture of what you are seeing.

rijotech commented:
> @rijotech please try
> ~/rpi-rgb-led-matrix/examples-api-use/demo --led-gpio-mapping=regular --led-rows=32 --led-cols=64 --led-row-addr-type=0 --led-chain=4 --led-show-refresh --led-slowdown-gpio=1 --led-parallel=3 --led-panel-type=FM6126A --led-pwm-bits=7 --led-pwm-dither-bits=1 --led-pixel-mapper=V-Mapper:Z -D0
> If that fails, please post picture of what you are seeing.
> Obviously, your led chain will be 4 for 3x4 panels (mine was 5)

Thanks a lot @marcmerlin I already figured out I cloned your repo and figure it how to make it work. Thanks to you both! and stay with this pandemic.

lehni commented:
@marcmerlin is there a way to reverse the sequence of the chains, so I can go right to left, bottom to top in my wiring?

hzeller commented:
could you use use the mirror or rotate mappers stacked on the v-mapper ?

marcmerlin commented:
Henner's answer is best, my code was indeed only written for one direction, but hopefully the transforms on top will sort you out

lehni commented:
I've reversed the cable routing instead, all good :) Thank you both for the quick feedback!
